Pretty serialized output via --output

dan-da opened the pull request:
Adds additional --output options: jsonpretty and yaml, eg:

```
OPTIONS:
    ...
    -o, --output <output_fmt>     Output data serialisation: [json, jsonpretty, yaml]
```

These serialization options make it much nicer to view structured data than raw/compact json, which remains the default.  All sub-commands use the new serialization options.  Some examples:

```
safe cat --output=jsonpretty safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc
[
  "safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc",
  {
    "/auth_and_connect.rs": {
      "created": "2019-11-21T08:41:23Z",
      "link": "safe://hbyyyynrmkqs4tn3bm9y57b7xr6tqaoca9e3insb9frkcs814h5r663pcy",
      "modified": "2019-11-21T08:41:23Z",
      "size": "3546",
      "type": "text/x-rust"
    },
    "/auth_daemon.rs": {
      "created": "2019-11-21T08:41:23Z",
      "link": "safe://hbyyyyda4ityjhafq6suqqdxkyc8c19ap6eyexbgczgajd3a5etaa4ijh3",
      "modified": "2019-11-21T08:41:23Z",
      "size": "13101",
      "type": "text/x-rust"
    }
  }
]
```

```
$ safe cat --output=yaml safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc 
---
- "safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc"
- /auth_and_connect.rs:
    created: "2019-11-21T08:41:23Z"
    link: "safe://hbyyyynrmkqs4tn3bm9y57b7xr6tqaoca9e3insb9frkcs814h5r663pcy"
    modified: "2019-11-21T08:41:23Z"
    size: "3546"
    type: text/x-rust
  /auth_daemon.rs:
    created: "2019-11-21T08:41:23Z"
    link: "safe://hbyyyyda4ityjhafq6suqqdxkyc8c19ap6eyexbgczgajd3a5etaa4ijh3"
    modified: "2019-11-21T08:41:23Z"
    size: "13101"
    type: text/x-rust
```

```
$ safe dog --output=jsonpretty safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc
[
  "safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc",
  {
    "data_type": "PublishedSeqAppendOnlyData",
    "type_tag": 1100,
    "version": 0,
    "xorname": "5b14f39cbdaacc592ee106259a7d3857a0b0893c3aef3b17c54f62a72a0c4e27"
  }
]
```
```
$ safe dog --output=json safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc
["safe://hnyynyiptjhhhzsicasjqhrdnmgu78bm4bcrj8o7q6qazai8sfj3kbt8nqbnc",{"data_type":"PublishedSeqAppendOnlyData","type_tag":1100,"version":0,"xorname":"5b14f39cbdaacc592ee106259a7d3857a0b0893c3aef3b17c54f62a72a0c4e27"}]
```
This last example was the only machine-readable way to serialize output before now and is not very human friendly.

#### Details
A new helper function is added: serialize_output() that takes a structured value and an OutputFmt enum, and serializes accordingly.  With this abstraction, we can easily add additional serializations supported by serde if desired.

This commit also fixes at least one json syntax error in dog.rs, where json was being manually constructed with strings and did not parse correctly. Now these values are constructed programmatically.

There is a new dependency on serde_yaml.

#### Testing
All existing test cases pass.  I have not yet written any new test cases specifically for the new output formats.

<!--
#### Thank you for contributing!

Please reference the issue(s) which this pull request addresses using [keywords](https://help.github.com/articles/closing-issues-using-keywords/) such as:

```
Resolves #452
Fixes #363
Closes #408
```

---

Provide QA team and reviewer steps to test the resolution.
For example:

```
QA:
Easiest way to test this PR would be to:
- Start authd
- login
- safe wallet insert <problem link>
- Expect to see an authentication error output

```

---

Commit messages should conform to the format:

```
<type>(<scope>): <description>

[optional body]

```

For example:

```
fix(auth): proper values returned on auth_decode_ipc_msg errors

  - Test case for authenticator error -208 IncompatibleMockStatus
  - Test case for authenticator error -103 when decoding share MData
    request for non-existent MData

```

Commit `type` can be one of:
**feat**: New feature
**fix**: Bug fix
**docs**: Documentation only changes
**style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
**refactor**: A code change that neither fixes a bug nor adds a feature
**perf**: A code change that improves performance
**test**: Adding missing tests
**chore**: Changes to the build process or auxiliary tools and libraries such as documentation generation
**revert**: Reverting a feature, fix, or commit which introduces a regression or new bug

Commit `scope` is open to any succinct term which indicates the effected feature or component.

See [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)

---
Write your description below this line: -->

dan-da commented:
btw, a note about yaml.  Sometimes json can be ugly when there are embedded quotes or backslash because everything gets escaped.   yaml format does not do that, and is more human friendly for such data while still structured and machine readable.  So I thought it is a nice option to have.

bochaco commented:
Also wondering if we really need the regular json output, the only benefit is that we save some space when it's used by apps, but i guess apps should be able to parse the prettyjson as well, so what are your thoughts about this only benefit we get from the shorter json, do we keep it?

dan-da commented on safe-cli/subcommands/helpers.rs:
```diff
@@ -86,3 +88,33 @@ pub fn parse_tx_id(src: &str) -> Result<u64, String> {
+    T: Serialize,
+{
+
+    if fmt == OutputFmt::JsonPretty {
```
done.

dan-da commented:
I addressed above comments except that I haven't added any test cases yet.  hopefully tonight.

I changed the --output options to json, jsoncompact, and yaml.   This way, *json* is now pretty and is the default ie for --json.   For now I left jsoncompact rather than dropping it entirely.

I'm not certain if any of the commands will ever return very large serialized datasets, eg 1Mb or much greater.   I could imagine a huge directory listing of tens of thousands of files for example.  In that case the jsoncompact format is more efficient to write/parse and saves space if written to disk.  So I would say it it is worth keeping but doesn't need to be the default.  Related to that, I noticed that serde also supports MessagePack which is "an efficient binary format that resembles a compact JSON" and some similar binary formats, so we might consider adding one or more of those -- now or in future.  See [supported formats list](https://serde.rs/#data-formats).

bochaco commented:
I think what you say makes sense chaging th default, so let's go with that, I think once we have the tests this should be ready to go, nice job @dan-da !

dan-da commented:
added tests for --output=jsoncompact and --output=yaml.   based on a test in cli_dog.rs

bochaco commented on safe-cli/tests/cli_dog.rs:
```diff
@@ -62,6 +62,98 @@ fn calling_safe_dog_files_container_nrsurl() {
+        .unwrap();
+
+    let content_info: (String, SafeData) = serde_json::from_str(&dog_output)
+        .expect("Failed to parse output of `safe dog` with -ii on file");
```
Minor thing, the error message is incorrect

dan-da replied:
changed to Failed to parse output of `safe dog`

bochaco commented on safe-cli/tests/cli_dog.rs:
```diff
@@ -62,6 +62,98 @@ fn calling_safe_dog_files_container_nrsurl() {
+        .unwrap();
+
+    let content_info: (String, SafeData) = serde_yaml::from_str(&dog_output)
+        .expect("Failed to parse output of `safe dog` with -ii on file");
```
Minor thing also about error message being innacurate

dan-da replied:
changed to Failed to parse output of `safe dog`

dan-da commented on safe-cli/subcommands/helpers.rs:
```diff
@@ -86,3 +88,22 @@ pub fn parse_tx_id(src: &str) -> Result<u64, String> {
+            .unwrap_or_else(|_| "Failed to serialise output to yaml".to_string()),
+        OutputFmt::Pretty => {
+            // For now we panic.  maybe make a default Pretty output later.
+            panic!("Pretty format must be handled by caller")
```
serialise_output is called in a lot of places as input to println(), eg: 

    println!("{}", serialise_output(&jsonv, output_fmt));

So that's why I wanted it to return a string and handle errors internally.  If it returns a Result<String,String>, how to handle the println() calls?

Or more generally, what is your recommendation for this situation?

re panic!() call, I could provide a default implementation for ::PRETTY, eg serialize it to pretty json.

re unwrap_or_else, I used it because I saw it being used in other places, ie:

```
$ grep -r unwrap * | grep -v tests
cli.rs:        let fmt = args.output_fmt.clone().unwrap_or_else(|| "".to_string());
operations/auth_daemon.rs:    let endpoint = notifs_endpoint.unwrap_or_else(|| AUTH_REQS_NOTIFS_ENDPOINT.to_string());
operations/auth_daemon.rs:    let endpoint = notifs_endpoint.unwrap_or_else(|| AUTH_REQS_NOTIFS_ENDPOINT.to_string());
operations/auth_daemon.rs:        .unwrap_or_else(|| String::from(""));
operations/auth_daemon.rs:        .unwrap_or_else(|| String::from(""));
subcommands/dog.rs:                    media_type.clone().unwrap_or_else(|| "Unknown".to_string())
subcommands/dog.rs:                        "media_type": media_type.clone().unwrap_or_else(|| "Unknown".to_string()),
subcommands/keys.rs:            let target = keyurl.unwrap_or_else(|| "".to_string());
subcommands/keys.rs:        let amount = preload.unwrap_or_else(|| PRELOAD_TESTCOINS_DEFAULT_AMOUNT.to_string());
subcommands/wallet.rs:                            let unwrapped_key_pair = key_generated_output
subcommands/wallet.rs:                            unwrapped_key_pair.sk
subcommands/helpers.rs:    let the_message = message.unwrap_or_else(|| "...awaiting data from STDIN stream...");
subcommands/helpers.rs:    let mut sk = sk.unwrap_or_else(|| String::from(""));
```

I'd like to do things the "right" way, just still learning rust, so not entirely sure what that is.

bochaco replied:
Yes, it's fine, all the places you are calling the `serialise_output` form already return a `Result`, that means you can call `serialise_output(...)?` from there, e.g. with println it will become `println!("{}", serialise_output(&jsonv, output_fmt)?);`. The `serialise_output` is unwrapped with `?` and it will resolve to the `String` returned if it was successful, or it will otherwise throw the error up exiting from the function you are calling `serialise_output` from. I'm not sure if I'm clear enough, but if you try to convert the return to be a Result as I suggested and then make sure you invoke it with a `?` at the end you'll be fine. Let me know if you want me to clarify it as I may not be that clear so, just ping me here

bochaco replied:
> re unwrap_or_else, I used it because I saw it being used in other places

I meant the `.unwrap()`, which I don't think there is any in your code, the `unwrap_or_else` is different,it won't panic, it let's you unwrap an `Option` and provide a default value if it founds a `None`. So `unwrap_or_else` is fine, but you shouldn't use `.unwrap()` on any non-test code as it means it could panic when unwrapping a `None` from an `Option`, or when unwrapping an `Err` from a `Result`.

dan-da replied:
well that sounds good except not all callers return a result.

```
   Compiling safe-cli v0.5.0 (/home/danda/dev/maidsafe/safe-api/safe-cli)
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
   --> safe-cli/subcommands/keys.rs:194:24
    |
194 |         println!("{}", serialise_output(&(&xorurl, pair), output_fmt)?);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
    |
    = help: the trait `std::ops::Try` is not implemented for `()`
    = note: required by `std::ops::Try::from_error`

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
   --> safe-cli/subcommands/keys.rs:196:24
    |
196 |         println!("{}", serialise_output(&xorurl, output_fmt)?);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
    |
    = help: the trait `std::ops::Try` is not implemented for `()`
    = note: required by `std::ops::Try::from_error`

error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
   --> safe-cli/subcommands/nrs.rs:161:13
    |
161 |             serialise_output(&(xorurl, processed_entries), output_fmt)?
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
    |
    = help: the trait `std::ops::Try` is not implemented for `()`
    = note: required by `std::ops::Try::from_error`

error: aborting due to 3 previous errors
```

As I see it, options are:

a. implement a default case for Pretty in serialise_json() and go back to returning a simple String.
b. handle the error in the calling functions, prior to calling println!()
c. modify the return type of calling functions and handle the error above them.

To me, a is simplest, and could look like:

```
// serialize structured value using any format from OutputFmt
pub fn serialise_output<T: ?Sized>(value: &T, fmt: OutputFmt) -> String
where
    T: Serialize,
{
    match fmt {
        OutputFmt::Yaml => serde_yaml::to_string(&value)
            .unwrap_or_else(|_| "Failed to serialise output to yaml".to_string()),
        OutputFmt::Json => serde_json::to_string_pretty(&value)
            .unwrap_or_else(|_| "Failed to serialise output to json".to_string()),
        OutputFmt::JsonCompact => serde_json::to_string(&value)
            .unwrap_or_else(|_| "Failed to serialise output to json".to_string()),
        OutputFmt::Pretty => {
            warn!("OutputFmt::Pretty' must be handled by caller, in serialise_output()");
            "OutputFmt::Pretty' must be handled by caller, in serialise_output()".to_string()
        }
    }
}
```
Alternatively, OutputFmt::Pretty could just serialize pretty json as a fallback.  And/or the warn!() could be an assert.

let me know what you think......

bochaco replied:
a. it could be ok for now since all the callers are simply sending the result directly to stdout and not doing anything if there is an error. In the future you may need to handle an error in different ways, ad that's when you'll need a Result.
b. that would be redundant code for verifying the valid inputs for serialise_output, which to me it should be part of the function itself, like any other function, validate its input and return an error if something is not accepted
c. this will be valid, but i think the functions which are giving you problems are the one that already print out the result to stdout, and that's why they don't return anything, they are expected to send the output to stdout which ever it is. So this option is also valid, those functions are the ones to decide what to do with the error as there is no other function expecting for its result, only the user is expecting for the result.

So, up to you, to me option a or c sounds ok.

bochaco replied:
As a general comment, functions like serialise_output are internal helper functions, so if they can fail they should return a Result, and then depending who the caller is would handle it in different ways, that's how you can decouple the logic from the caller and the function. If one day I want to call your function and send the succesfull output through a USB device, but send the error through a lo file, I'll need to be able to know if it's an error or a sucessfull result, and this is when a `Result` comes in. A helper function shouldn't make any type of assumption of how the returned value is going to be used by the caller, thus it should tell you "everything went ok, this is the result: <outcome>", or "we had a problem, which is: <problem>"

dan-da replied:
Ok, I decided to go with a for now, as it is simplest.  If you'd like to modify as you think best, perhaps I can learn from it.  Otherwise, I'm happy enough with it.

One possible improvement I thought of would be to accept a closure param (pretty_serializer) that would be called to write the OutputFmt::Pretty case.  But would need to be optional somehow for now, because it's a big change to modify all the places where Pretty output is generated now.

bochaco replied:
I think it's good enough for now, I personally prefer to evolve it as it's needed, so option a fits in current needs, if in the future we need to handle that error case in different ways from a caller we then introduce a Result and do what's necessary to other callers, so all good.

bochaco commented:
Thanks a lot @dan-da . Now, for next PRs please make sure you follow the convention we have for our commits, look at this here: https://github.com/maidsafe/safe-api#commits-and-pull-requests and look at some of our prevous commits to get an idea, otherwise we'll have to squash your commits into a single one, which is also an option, so just keep it in mind for next one.

dan-da commented:
re commit conventions.  ok, got it.  thx for pointer to docs.
